Optimize CUP parser: replace legacy collections, add AST features

EmptyDreams opened the pull request:
This changeset improves code performance, introduces AST automation, but may require compatibility checks for generated parser code.

Below is a brief summary of the pull request content:

1. **Collection Optimization**
  - Replaced `java.util.Stack` with `ArrayStack/IntArrayStack`
  - Replaced `Hashtable` with `HashMap` for better performance  
  - Replaced `StringBuffer` with `StringBuilder` for efficient string operations  
  - Optimized integer-keyed maps with `MonotonicIntObjectArrayMap` (leveraging strictly increasing keys for O(1) insertion and O(log n) lookup)  
  - Replaced string concatenation in loops with `StringBuilder`  

2. **AST Automation**  
  - Added automatic AST node class generation based on `.cup` files 
  - Implemented AST flattening for list structures (e.g., converting recursive nodes to `List<T>`) 
  - Introduced `-ast` and `-ast_flatten` parameters

3. **Code Structure & Performance**  
  - Removed redundant code (e.g., always-true boolean expressions)  
  - Refactored code style

4. **Compatibility & Testing**  
  - Modified `do_action` method signature in `lr_parser` (potential backward compatibility impact)  
  - Auto-generate output directories if missing  
  - Recommended thorough testing despite passing initial local checks

EmptyDreams committed:
Refactor: Optimize code structure and performance

+ Replace Stack with ArrayStack, and Hashtable with HashMap
+ Optimize some code logic to improve execution efficiency
+ Modify code formatting to enhance readability

EmptyDreams committed:
revert edit

EmptyDreams committed:
refactor: Replace `ArrayStack<Integer>` with `IntArrayStack`

EmptyDreams commented:
Apart from the contents of the `java_cup` package, I haven't modified most of the files.  

At the same time, I kept `Map<Integer, *>`. `IntMap` is relatively more complex, and using other open-source implementations would be simpler. However, this involves open-source licensing, and I'm not entirely sure which library to choose. So, I've left it as is for now.

EmptyDreams committed:
refactor: Optimize the code structure and style of the Main class.

EmptyDreams committed:
refactor: Use MonotonicIntObjectArrayMap to optimize integer key-value storage

EmptyDreams committed:
docs: Add comments

EmptyDreams commented:
I replaced `Map<Integer, *>` with `MonotonicIntObjectArrayMap<*>`. For detailed information about this class, please refer to its documentation comments.

I noticed that the keys in `Map<Integer, *>` follow a strictly increasing pattern in the code. I believe using an array (with O(1) insertion performance, except during space expansion) combined with binary search (O(log n) lookup performance) would be an acceptable alternative. Therefore, I've provided a simplified implementation.

If you find this implementation unsatisfactory, feel free to modify or remove the code as needed.

EmptyDreams committed:
feat: Implement automatic AST generation with a new control parameter (-ast)

EmptyDreams commented:
I made further modifications to the code. Now, CUP can automatically generate AST node classes based on the content of `.cup` files, making subsequent syntax analysis more convenient.  

Additionally, I added a new variable "name" to the code context of `.cup` nonterminals to indicate the current node's name. This change primarily avoids redundant name encoding in `.cup` files.  

In the next step, I plan to further improve the automation of AST generation, allowing it to proceed without manually writing code blocks for nonterminals.

EmptyDreams committed:
fix: fix enforced addition of 'Node' suffix in AstNode class name

EmptyDreams committed:
fix: Prohibit the same label from corresponding to multiple types.

EmptyDreams committed:
Support auto-generation of code for building AST nodes and fix errors occurring in certain cases.

EmptyDreams committed:
fix: Fix resizing logic in IntArrayStack

EmptyDreams commented:
I've made further optimizations to the code and addressed several issues from the previous commit:

1. Fixed incorrect class name generation for AstNode classes where they erroneously ended with "Node"
2. Resolved an issue where using the same label with different types across production rules for the same non-terminal would generate incorrect results (the solution was to simply throw an error in such cases - I believe supporting this scenario isn't necessary as it would overcomplicate AST processing code)
3. Patched a bug in `IntArrayStack` where incorrect expansion logic could lead to array index out-of-bounds exceptions

In this update, I've implemented:
- Automated AST construction and automatic skipping of empty nodes

EmptyDreams committed:
Auto-generate output directory if it doesn't exist, instead of throwing an exception.

DrMichaelPetter commented:
Thanks for all the effort, I am currently traveling without a computer, so can not review ATM. I will reach out, when I am back.

EmptyDreams committed:
Flatten AST nodes and fix some bugs

EmptyDreams commented:
In this submission, I fixed some bugs in the code I wrote earlier.  

Additionally, I implemented flattening for AST nodes. Previously, for nodes like `Xxx*` or `Xxx+`, the generated structure was a tree with a depth equal to the length of the list, which was quite inconvenient to use. I added some code to automatically convert such structures into an `List`.  

Here’s the specific transformation logic:  

1. I added a parameter `-ast_flatten <suffix>`, which specifies a fixed suffix. If a nonterminal’s name ends with this suffix, I will attempt to determine whether it represents a list (including nullable lists `Xxx*` and non-nullable lists `Xxx+`).  
2. If condition 1 is met, I will check whether the nonterminal has two or three productions.  
3. If the nonterminal has two productions and meets the following conditions, it is considered a non-nullable list:
   - One production is another nonterminal or terminal (`T`).
   - One production is `self T` (where `self` refers to the current nonterminal itself).
   - `self` does not carry a label, while `T` does.
4. If the nonterminal has three productions and meets the following conditions, it is identified as a nullable list:  
   - One production is an empty expression (indirect empty expressions are currently unsupported).  
   - One production is another nonterminal or terminal (`T`).  
   - One production is `self T`.  
   - `self` does not carry a label, while `T` does.  

Once a nonterminal is confirmed to be a list, its type will be converted from `AstNode` to `List<T>`. (This direct type change is intended to reduce tree depth—if `AstNode` were retained, it would introduce an additional layer.)  

Here’s how I define a **non-nullable list** (omitting nonterminal/terminal declarations and precedence rules):  

```cup
NT_ITEM ::= T_TEST0 T_VALUE:value;
NT_ITEM_LIST ::= NT_ITEM:item | NT_ITEM_LIST NT_ITEM:item;
```

And for a **nullable list**:  

```cup
NT_ITEM ::= T_TEST0 T_VALUE:value;
NT_ITEM_LIST ::= NT_ITEM:item | NT_ITEM_LIST NT_ITEM:item | /* empty */;
```

Note: The aforementioned `Xxx*` and `Xxx+` are analogies—CUP currently does not support this syntax.

EmptyDreams committed:
AST flattening forces specific-suffix non-terminals into lists

EmptyDreams commented:
In this commit, I made the list expression check for `ast_flatten` mandatory.  

Previously, if a non-terminal's name ended with the suffix set by `ast_flatten` but wasn't actually a list, it would be treated as a regular node.  

However, upon reconsideration, I found this design problematic—it could easily lead to hard-to-detect issues in `.cup` files.  

Therefore, I changed it so that if a non-terminal's name matches the configured suffix but its production doesn't meet list requirements, an exception is now thrown directly.

EmptyDreams committed:
refactor code and fix bug

EmptyDreams commented:
In this round of modifications, I made the following changes:  

1. Relaxed the restrictions on list expressions. Now, for each list-type non-terminal, it only needs to meet the following conditions:  
   + There must be exactly one production that contains only one symbol, and this symbol must have a label.  
   + There may be zero or one empty production.  
   + There must be one or more recursive productions that reference itself while also referencing list elements (both the self-reference and the list element reference must occur exactly once). The productions may contain other symbols, but except for the list element, no other symbols are allowed to have labels.  

2. Renamed `AstNode` to `IAstNode` and moved it from being generated at runtime to being placed in the `runtime` package.  

3. Added generic type parameters for list types.  

4. Fixed an issue where the inserted `name` variable could easily conflict with user-defined `label`s.  

5. Fixed an issue where the generated AST node classes might contain conflicting constructors.

EmptyDreams committed:
style: Remove unnecessary blank lines

EmptyDreams committed:
Fix potential duplication of class member variables for AST nodes

EmptyDreams committed:
feat: Support special handling for boolean flags

EmptyDreams committed:
Verify the validity of existence-checking labels during AST construction.

EmptyDreams commented:
For automatically generated ASTs, I've added support for symbol existence checking:  

**"When a label matches any of the patterns `is_.*`, `is[A-Z].*`, or `is[0-9].*`, I classify it as an existence-checking label."**  

Existence-checking labels do not store the actual `symbol` value in the AST node—only a boolean flag indicating whether the symbol exists.  

### Why this matters:  
When working with ASTs, there are cases where we need to determine which production rule (or any matching rule) generated a node. This feature allows such checks by testing whether a symbol exists.  

### Design rationale:  
Existence-checking labels are explicitly separated from ordinary labels to:  
1. Maintain a clear distinction(e.g., checking presence without caring about the value), and  
2. Handle cases where AST inspection relies on untyped terminals.

EmptyDreams committed:
refactor: Further relax the restrictions on list expressions

- No longer require a production with only a single element to exist necessarily.

EmptyDreams committed:
feat: Support the use of indirect empty expressions in list expressions.

EmptyDreams committed:
fix bug

EmptyDreams commented:
In this commit, I made the following changes:

1. Fixed some bugs
2. Further relaxed restrictions on list expressions by no longer requiring a production containing only elements, to support more types of list expressions
3. Added support for using indirect empty nodes to represent `/* empty */` in list expressions:
    For example:
    ```cup
    nonterminal empty;
    empty ::= /* empty */;
    ```
    When using `empty` in lists, it will be correctly recognized as an empty node.

---

The number of commits and additional comments I've submitted might be relatively large, as I was making code modifications while experimenting and testing the changes myself. I sincerely apologize if this has caused any inconvenience.

EmptyDreams committed:
Fix the issue where the generated AST existence check has different setter calls and definition signatures.

EmptyDreams committed:
Update the binary file

EmptyDreams commented:
I'm closing this PR for now as I still need to make some changes. I'll submit a new one later with all modifications finalized, along with detailed documentation on the changes and new features. Feel free to check my repo for updates or discussion. :)

https://github.com/EmptyDreams/cup
